lib/kvdb/sched_sts.c

```diff
@@ -1,473 +1,84 @@
-                         (ulong)dbg_wake_timeouts,
-                         (ulong)dbg_wake_other,
-                         ss);
-
-                dbg_wake_signals = 0;
-                dbg_wake_timeouts = 0;
-                dbg_wake_other = 0;
-                dbg_last = now;
-            }
-        }
-
-        if (ss == SS_EXIT)
-            break;
-
-        timeout_ms = MONITOR_TIMEOUT_MS;
-
-        if (add_worker_next_ns <= get_time_ns()) {
-            merr_t err;
-            uint   i;
-
-            for (i = 0; i <= self->qc; i++) {
-                if (sts_worker_surplus(self, i) < 0) {
-                    err = add_worker(self, i);
-                    if (!err) {
-                        add_worker_next_ns = 0;
-                        timeout_ms = 10;
-                    } else {
-                        add_worker_next_ns = get_time_ns() + 5 * NSEC_PER_SEC;
-                    }
-                }
-            }
-        }
-    }
-
-    if (csched_rp_dbg_mon(self->rp))
-        log_info("sts/mon exiting, waiting for %d workers to finish",
-                 atomic_read(&self->spawned_workers));
-
-    while (atomic_read(&self->spawned_workers) > 0)
-        msleep(EXIT_SLEEP_MS);
-
-    cancel_jobs_all(self);
-
-    return 0;
-}
-
-static void
-sts_perfc_alloc_internal(struct sts *self)
+sts_job_submit(struct sts *self, struct sts_job *job)
 {
-    char namebuf[32], qnum_str[16];
-    int  i;
-
-    snprintf(namebuf, sizeof(namebuf), "kvdb/%s", self->name);
-
-    /* Iterate w/ <= to get shared worker stats */
-    for (i = 0; i <= self->qc; i++) {
-        snprintf(qnum_str, sizeof(qnum_str), "q%d", i);
-        sts_perfc_alloc(self->rp->perfc_level, namebuf, qnum_str, &self->qv[i].qpc);
-    }
-}
+    assert(self && job);
+    assert(job->sj_job_fn);
 
-static void
-sts_perfc_free_internal(struct sts *self)
-{
-    int i;
+    if (csched_rp_dbg_jobs(self->sts_rp))
+        log_info("sts/job %u submit qnum %u", job->sj_id, job->sj_qnum);
 
-    /* Iterate w/ <= to get shared worker stats */
-    for (i = 0; i <= self->qc; i++)
-        sts_perfc_free(&self->qv[i].qpc);
+    INIT_WORK(&job->sj_work, sts_job_run);
+    queue_work(self->sts_wq, &job->sj_work);
 }
 
 merr_t
 sts_create(struct kvdb_rparams *rp, const char *name, uint nq, struct sts **handle)
 {
-    int         rc;
-    uint        i, len;
     struct sts *self;
 
     assert(rp);
     assert(name);
     assert(nq);
     assert(handle);
 
-    if (ev(nq > STS_QUEUES_MAX))
-        return merr(EINVAL);
+    *handle = NULL;
 
-    self = alloc_aligned(sizeof(*self), HSE_L1D_LINESIZE);
+    self = aligned_alloc(alignof(*self), sizeof(*self));
     if (ev(!self))
         return merr(ENOMEM);
 
     memset(self, 0, sizeof(*self));
+    self->sts_rp = rp;
 
-    /* Name pointer is safe -- it will not go away */
-    self->name = name;
+    snprintf(self->sts_name, sizeof(self->sts_name), "hse_sts_%u_%s",
+             rp->csched_policy, name);
```

Comment:
I don't see any value having `csched_policy` being part of the name.

In the old days we different implementations of the scheduler (sp1, sp2 ).  Now all we have is sp3 and noop.  The policy is used to pick an implementation and and populate a function ptr ops struct.  I would simplify as follows:
1. Bring "sp3" into "csched" and lose the concept of "sp3"
2. Eliminate the concept of a csched policy as it currently exists.
3. Eliminate the function ops struct.
4. Figure out another way to disable compaction instead of using the "noop" implementation.

A future PR perhaps?

Reply:
Yeah, lemme open another PR for that..

Reply:
I've got another commit coming that gets rid of the bonus stuff, keeps us from calling sp3_work 8000+ times/sec when there are a lot of trees, and other little schduling related issues...